source/opt/loop_fission.cpp

```diff
@@ -0,0 +1,509 @@
+  std::set<ir::Instruction*> seen_instructions_;
+
+  // A map of instructions to their relative position in the function.
+  std::map<ir::Instruction*, size_t> instruction_order_;
+
+  ir::IRContext* context_;
+
+  ir::Loop* loop_;
+
+  // This is set to true by TraverseUseDef when traversing the instructions
+  // related to the loop condition and any if conditions should any of those
+  // instructions be a load.
+  bool load_used_in_condition_;
+};
+
+bool LoopFissionImpl::MovableInstruction(const ir::Instruction& inst) const {
+  return inst.opcode() == SpvOp::SpvOpLoad ||
+         inst.opcode() == SpvOp::SpvOpStore ||
+         inst.opcode() == SpvOp::SpvOpSelectionMerge ||
+         inst.opcode() == SpvOp::SpvOpPhi || inst.IsOpcodeCodeMotionSafe();
+}
+
+void LoopFissionImpl::TraverseUseDef(ir::Instruction* inst,
+                                     std::set<ir::Instruction*>* returned_set,
+                                     bool ignore_phi_users, bool report_loads) {
+  assert(returned_set && "Set to be returned cannot be null.");
+
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+  std::set<ir::Instruction*>& inst_set = *returned_set;
+
+  // We create this functor to traverse the use def chain to build the
+  // grouping of related instructions. The lambda captures the std::function
+  // to allow it to recurse.
+  std::function<void(ir::Instruction*)> traverser_functor;
+  traverser_functor = [this, def_use, &inst_set, &traverser_functor,
+                       ignore_phi_users, report_loads](ir::Instruction* user) {
+    // If we've seen the instruction before or it is not inside the loop end the
+    // traversal.
+    if (!user || seen_instructions_.count(user) != 0 ||
+        !context_->get_instr_block(user) ||
+        !loop_->IsInsideLoop(context_->get_instr_block(user))) {
+      return;
+    }
+
+    // Don't include labels or loop merge instructions in the instruction sets.
+    // Including them would mean we group instructions related only by using the
+    // same labels (i.e phis). We already preempt the inclusion of
+    // OpSelectionMerge by adding related instructions to the seen_instructions_
+    // set.
+    if (user->opcode() == SpvOp::SpvOpLoopMerge ||
+        user->opcode() == SpvOp::SpvOpLabel)
+      return;
+
+    // If the |report_loads| flag is set, set the class field
+    // load_used_in_condition_ to false. This is used to check that none of the
+    // condition checks in the loop rely on loads.
+    if (user->opcode() == SpvOp::SpvOpLoad && report_loads) {
+      load_used_in_condition_ = true;
+    }
+
+    // Add the instruction to the set of instructions already seen, this breaks
+    // recursion and allows us to ignore certain instructions.
+    seen_instructions_.insert(user);
+
+    // We want to traverse the operands of branches but we don't want to include
+    // them in the set.
+    if (!user->IsBranch()) {
+      inst_set.insert(user);
+    }
```

Comment:
But if the branch itself is not included, how is this going to help?  I'm missing something here.  ISTM that you'd be splitting the branches of the conditional, but not the predicate itself.

Reply:
When we group the instructions we first loop over the conditional control flow instructions and call each instruction with this traverse function to ensure that it is added to the seen_instructions_ set. We just ignore the inst_sets produced by that pass but it will mean that any, say, OpBranchConditionals and associated instructions will be ignored by the actual grouping pass which creates the sets. 

This check was to then stop any unconditional branches which would not be part of the effective blacklist created by the seen_instructions_ set. Looking at it now, I think it's confusing to break up the logic like that so I've moved the check to the same part that we deal with conditional branches.